corehq/apps/ivr/migrations/0002_call_app_id.py

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.22 on 2019-10-28 20:52
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+
+from corehq.apps.ivr.models import Call
+from corehq.apps.app_manager.dbaccessors import get_apps_in_domain
+from corehq.util.django_migrations import skip_on_fresh_install
+
+
+app_id_by_form_unique_id = {}
+
+
+def _update_model(model, domain):
+    if model.form_unique_id is None:
+        return None
+
+    if model.form_unique_id not in app_id_by_form_unique_id:
+        apps = get_apps_in_domain(domain)
+        for app in apps:
+            for module in app.modules:
+                for form in module.get_forms():
+                    app_id_by_form_unique_id[form.unique_id] = app.get_id
+
+    model.app_id = app_id_by_form_unique_id.get(model.form_unique_id, None)
+    if model.app_id:
+        model.save()
+
+
+@skip_on_fresh_install
+def _populate_app_id(apps, schema_editor):
+    for call in Call.objects.filter(form_unique_id__isnull=False).all():
```

Comment:
Do any of these models have enough rows that the migrations should fetch them in chunks?

Reply:
Yes, `Call` and the two event models both have hundreds of thousands of rows (there are counts in the description for https://github.com/dimagi/commcare-hq/pull/25757). I thought `all` was reasonably performant, is it not? Added transactions in a8909676718 to speed up the saves.

Reply:
>  I thought `all` was reasonably performant, is it not?

The issue with `all` is that it pulls the entire table into memory, not that it would be slower than pulling in chunks.

Also transactions do not generally make things faster (actually should make them slower). They are only intended to ensure that either all changes happen or do not. A bulk update would be more efficient `Call.objects.filter(form_unique_id=unique_id).update(app_id=app_id)`

The most effecient migration here will probably look something like:

```python
Call.objects.distinct(form_unique_id) # or however you should write it https://docs.djangoproject.com/en/dev/ref/models/querysets/#django.db.models.query.QuerySet.distinct
# Generate the form_unqiue_id -> app_id dictionary
for id in form_unique_id:
    Call.objects.filter(form_unique_id=unique_id).update(app_id=app_id)
```

Reply:
Thanks for the suggestions. [This](https://stackoverflow.com/questions/3395236/aggregating-saves-in-django) is what led me to add transactions would improve performance.

Reply:
Ah. that makes sense, and I do believe that it would make it faster in sqlite because it doesn't have the same concurrency model as postgres (/ doesn't have a concurrency model). Later in that question someone commented about why its faster than sqllite https://stackoverflow.com/questions/3395236/aggregating-saves-in-django#comment92587635_3395488 and basically says its due to happening in the same file lock/unlock process.

From my understanding there's more overhead to long running transactions in postgres because it needs to isolate commits that happen in transactions from other concurrent queries (https://www.postgresql.org/docs/9.6/transaction-iso.html). It likely wouldn't have had a real impact on this PR, but wanted to guard again the thought that transactions == better performance.